# Stage 1: Enhanced Data Cleaning, Preprocessing, and Exploratory Analysis


## 1. Data Collection and Cleaning


### Overview of JSON Parsing Issues with Initial Method:

Original Method:

- Initially attempted to load a JSON file using `pandas.read_json`, which resulted in errors due to file content issues.

Issues and Solutions:

1. **Trailing Data Error**:

   - **Problem**: Encountered a `ValueError` related to "trailing data" using `pandas.read_json` .
   - **Solution**: Inspected the JSON file for proper formatting and structure, ensuring no unexpected data followed the JSON content.

2. **Character Encoding Error**:

   - **Problem**: After resolving the first issue, a `UnicodeDecodeError` was encountered, suggesting incompatible character encoding.
   - **Solution**: Modified the file-reading command to include `encoding='utf-8'` , supporting a broader range of characters.

3. **Multiple JSON Objects Error**:

   - **Problem**: Following the encoding fix, a `JSONDecodeError` due to "extra data" indicated multiple JSON objects concatenated without proper formatting.
   - **Solution**:

     - If each line is a separate JSON object, parsed each line individually.
     - For multiple concatenated objects in a single stream, manually split and restructured the data for proper parsing.

The steps implemented below address these issues, enhancing the robustness of JSON data handling and parsing in diverse file structures.


In [22]:
import json
import pandas as pd

data = []

# Read the JSON file line by line, assuming each line is a complete JSON object
with open("patent_data.json", "r", encoding="utf-8") as file:
    for line in file:
        # Convert each line to a Python dictionary
        json_obj = json.loads(line)
        data.append(json_obj)

# Convert the list of dictionaries to a DataFrame
cleantech_patent_data = pd.DataFrame(data)

# Print the shape of the cleantech_patent_data dataframe
print(cleantech_patent_data.shape)

# Display the cleantech_patent_data dataframe
cleantech_patent_data

(30000, 8)


publication_number application_number country_code  \
0       US-2022239235-A1  US-202217717397-A           US   
1       US-2022239251-A1  US-202217580956-A           US   
2          EP-4033090-A1      EP-21152924-A           EP   
3          EP-4033090-A1      EP-21152924-A           EP   
4         US-11396827-B2  US-202117606042-A           US   
...                  ...                ...          ...   
29995     CN-214835218-U  CN-202121309411-U           CN   
29996     CN-113690940-A  CN-202111107806-A           CN   
29997     CN-113683210-A  CN-202110998828-A           CN   
29998     CN-113683210-A  CN-202110998828-A           CN   
29999     CN-113682495-A  CN-202110997949-A           CN   

                                         title_localized  \
0      [{'text': 'Adaptable DC-AC Inverter Drive Syst...   
1      [{'text': 'System for providing the energy fro...   
2      [{'text': 'Verfahren zum steuern einer windene...   
3      [{'text': 'Verfahren zum steuern einer windene...   
4      [{'text': 'Control method for optimizing solar...   
...                                                  ...   
29995  [{'text': '一种具有充电功能的无人机机库', 'language': 'zh', ...   
29996  [{'text': '一种风电供电控制方法', 'language': 'zh', 'tru...   
29997  [{'text': '一种基于无线控制的漂浮式太阳能曝气装置', 'language': '...   
29998  [{'text': '一种基于无线控制的漂浮式太阳能曝气装置', 'language': '...   
29999  [{'text': 'Space film satellite integrating co...   

                                      abstract_localized publication_date  \
0      [{'text': 'Disclosed is an adaptable DC-AC inv...         20220728   
1      [{'text': 'In accordance with an example embod...         20220728   
2      [{'text': 'Verfahren zum Steuern einer Windene...         20220727   
3      [{'text': 'Verfahren zum Steuern einer Windene...         20220727   
4      [{'text': 'A control method for optimizing a s...         20220726   
...                                                  ...              ...   
29995  [{'text': 'The utility model relates to an unm...         20211123   
29996  [{'text': 'The invention provides a wind power...         20211123   
29997  [{'text': '本发明公开了一种基于无线控制的漂浮式太阳能曝气装置，包括漂浮板、太阳能...         20211123   
29998  [{'text': '本发明公开了一种基于无线控制的漂浮式太阳能曝气装置，包括漂浮板、太阳能...         20211123   
29999  [{'text': 'The invention discloses a space fil...         20211123   

                                                inventor  \
0                                                     []   
1                                                     []   
2      [Schaper, Ulf, von Aswege, Enno, Gerke Funcke,...   
3      [Schaper, Ulf, von Aswege, Enno, Gerke Funcke,...   
4                                                     []   
...                                                  ...   
29995  [ZHANG ANZHI, ZHANG WANYONG, HUANG HUIYONG, XU...   
29996         [ZHANG YUCHUAN, LONG HAIYANG, TAN CUNZHEN]   
29997                                        [An Zhixia]   
29998                                        [An Zhixia]   
29999  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...   

                                                     cpc  
0      [{'code': 'H02M7/5395', 'inventive': True, 'fi...  
1      [{'code': 'H02S40/38', 'inventive': True, 'fir...  
2      [{'code': 'F03D7/0276', 'inventive': True, 'fi...  
3      [{'code': 'F03D7/0276', 'inventive': True, 'fi...  
4      [{'code': 'F24S50/00', 'inventive': True, 'fir...  
...                                                  ...  
29995  [{'code': 'Y02E10/50', 'inventive': False, 'fi...  
29996  [{'code': 'Y02E10/76', 'inventive': False, 'fi...  
29997  [{'code': 'C02F2209/22', 'inventive': False, '...  
29998  [{'code': 'C02F2209/22', 'inventive': False, '...  
29999  [{'code': 'B64G4/00', 'inventive': True, 'firs...  

[30000 rows x 8 columns]

In [23]:
# Check for missing values

cleantech_patent_data.isnull().sum()

publication_number    0
application_number    0
country_code          0
title_localized       0
abstract_localized    0
publication_date      0
inventor              0
cpc                   0
dtype: int64

In [24]:
# Inspect the datatype of each columns first element

for col in cleantech_patent_data.columns:
    print(
        col,
        ": ",
        type(cleantech_patent_data[col].iloc[0]),
        cleantech_patent_data[col].iloc[0],
    )

publication_number :  <class 'str'> US-2022239235-A1
application_number :  <class 'str'> US-202217717397-A
country_code :  <class 'str'> US
title_localized :  <class 'list'> [{'text': 'Adaptable DC-AC Inverter Drive System and Operation', 'language': 'en', 'truncated': False}]
abstract_localized :  <class 'list'> [{'text': 'Disclosed is an adaptable DC-AC inverter system and its operation. The system includes multiple DC input sources as input to provide a stable operation under various conditions. DC input sources may be added to the system or removed from the system without impacting the functionality of the system. The disclosed system is suited for solar energy harvesting in grid-connected or off-grid modes of operation.', 'language': 'en', 'truncated': False}]
publication_date :  <class 'str'> 20220728
inventor :  <class 'list'> []
cpc :  <class 'list'> [{'code': 'H02M7/5395', 'inventive': True, 'first': False, 'tree': []}, {'code': 'H02J3/32', 'inventive': False, 'first': False, 

This step involves normalizing the nested JSON structures. Given the nature of the patent data, with nested fields such as 'title_localized' and 'abstract_localized', we will extract these nested dictionaries and flatten them to create a more accessible DataFrame format.

- Normalize Nested Structures: Since fields like 'title_localized' and 'abstract_localized' contain key details in nested dictionaries, we'll extract these details into separate DataFrame columns. This will facilitate easier access and manipulation of these fields in subsequent analyses.


In [25]:
# Extracting data from nested fields
cleantech_patent_data["publication_number"] = [
    entry["publication_number"] for entry in data
]
cleantech_patent_data["application_number"] = [
    entry["application_number"] for entry in data
]
cleantech_patent_data["country_code"] = [entry["country_code"] for entry in data]
cleantech_patent_data["title"] = [
    entry["title_localized"][0]["text"] if entry["title_localized"] else None
    for entry in data
]
cleantech_patent_data["abstract"] = [
    entry["abstract_localized"][0]["text"] if entry["abstract_localized"] else None
    for entry in data
]
cleantech_patent_data["publication_date"] = [
    entry["publication_date"] for entry in data
]
cleantech_patent_data["inventors"] = [entry["inventor"] for entry in data]

# Display the cleantech_patent_data dataframe
cleantech_patent_data

publication_number application_number country_code  \
0       US-2022239235-A1  US-202217717397-A           US   
1       US-2022239251-A1  US-202217580956-A           US   
2          EP-4033090-A1      EP-21152924-A           EP   
3          EP-4033090-A1      EP-21152924-A           EP   
4         US-11396827-B2  US-202117606042-A           US   
...                  ...                ...          ...   
29995     CN-214835218-U  CN-202121309411-U           CN   
29996     CN-113690940-A  CN-202111107806-A           CN   
29997     CN-113683210-A  CN-202110998828-A           CN   
29998     CN-113683210-A  CN-202110998828-A           CN   
29999     CN-113682495-A  CN-202110997949-A           CN   

                                         title_localized  \
0      [{'text': 'Adaptable DC-AC Inverter Drive Syst...   
1      [{'text': 'System for providing the energy fro...   
2      [{'text': 'Verfahren zum steuern einer windene...   
3      [{'text': 'Verfahren zum steuern einer windene...   
4      [{'text': 'Control method for optimizing solar...   
...                                                  ...   
29995  [{'text': '一种具有充电功能的无人机机库', 'language': 'zh', ...   
29996  [{'text': '一种风电供电控制方法', 'language': 'zh', 'tru...   
29997  [{'text': '一种基于无线控制的漂浮式太阳能曝气装置', 'language': '...   
29998  [{'text': '一种基于无线控制的漂浮式太阳能曝气装置', 'language': '...   
29999  [{'text': 'Space film satellite integrating co...   

                                      abstract_localized publication_date  \
0      [{'text': 'Disclosed is an adaptable DC-AC inv...         20220728   
1      [{'text': 'In accordance with an example embod...         20220728   
2      [{'text': 'Verfahren zum Steuern einer Windene...         20220727   
3      [{'text': 'Verfahren zum Steuern einer Windene...         20220727   
4      [{'text': 'A control method for optimizing a s...         20220726   
...                                                  ...              ...   
29995  [{'text': 'The utility model relates to an unm...         20211123   
29996  [{'text': 'The invention provides a wind power...         20211123   
29997  [{'text': '本发明公开了一种基于无线控制的漂浮式太阳能曝气装置，包括漂浮板、太阳能...         20211123   
29998  [{'text': '本发明公开了一种基于无线控制的漂浮式太阳能曝气装置，包括漂浮板、太阳能...         20211123   
29999  [{'text': 'The invention discloses a space fil...         20211123   

                                                inventor  \
0                                                     []   
1                                                     []   
2      [Schaper, Ulf, von Aswege, Enno, Gerke Funcke,...   
3      [Schaper, Ulf, von Aswege, Enno, Gerke Funcke,...   
4                                                     []   
...                                                  ...   
29995  [ZHANG ANZHI, ZHANG WANYONG, HUANG HUIYONG, XU...   
29996         [ZHANG YUCHUAN, LONG HAIYANG, TAN CUNZHEN]   
29997                                        [An Zhixia]   
29998                                        [An Zhixia]   
29999  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...   

                                                     cpc  \
0      [{'code': 'H02M7/5395', 'inventive': True, 'fi...   
1      [{'code': 'H02S40/38', 'inventive': True, 'fir...   
2      [{'code': 'F03D7/0276', 'inventive': True, 'fi...   
3      [{'code': 'F03D7/0276', 'inventive': True, 'fi...   
4      [{'code': 'F24S50/00', 'inventive': True, 'fir...   
...                                                  ...   
29995  [{'code': 'Y02E10/50', 'inventive': False, 'fi...   
29996  [{'code': 'Y02E10/76', 'inventive': False, 'fi...   
29997  [{'code': 'C02F2209/22', 'inventive': False, '...   
29998  [{'code': 'C02F2209/22', 'inventive': False, '...   
29999  [{'code': 'B64G4/00', 'inventive': True, 'firs...   

                                                   title  \
0      Adaptable DC-AC Inverter Drive System and Oper...   
1      System for providing the energy from a single ...   
2          Verfah

In [26]:
# Inspect the datatype of each columns of cleantech_patent_data's first element

for col in cleantech_patent_data.columns:
    print(
        col,
        ": ",
        type(cleantech_patent_data[col].iloc[0]),
        cleantech_patent_data[col].iloc[0],
    )

publication_number :  <class 'str'> US-2022239235-A1
application_number :  <class 'str'> US-202217717397-A
country_code :  <class 'str'> US
title_localized :  <class 'list'> [{'text': 'Adaptable DC-AC Inverter Drive System and Operation', 'language': 'en', 'truncated': False}]
abstract_localized :  <class 'list'> [{'text': 'Disclosed is an adaptable DC-AC inverter system and its operation. The system includes multiple DC input sources as input to provide a stable operation under various conditions. DC input sources may be added to the system or removed from the system without impacting the functionality of the system. The disclosed system is suited for solar energy harvesting in grid-connected or off-grid modes of operation.', 'language': 'en', 'truncated': False}]
publication_date :  <class 'str'> 20220728
inventor :  <class 'list'> []
cpc :  <class 'list'> [{'code': 'H02M7/5395', 'inventive': True, 'first': False, 'tree': []}, {'code': 'H02J3/32', 'inventive': False, 'first': False, 

We have successfully extracted the nested data into separate columns. Now, we will process the nested data in the 'cpc' column.


In [27]:
# For the 'cpc' column which is a list of dictionaries, extract it into a separate DataFrame and then process as required
cpc_data = pd.json_normalize(
    data, record_path=["cpc"], meta=["publication_number"], errors="ignore"
)

# Display the cpc_data dataframe

cpc_data

code  inventive  first tree publication_number
0        H02M7/5395       True  False   []   US-2022239235-A1
1          H02J3/32      False  False   []   US-2022239235-A1
2          H02M1/32       True  False   []   US-2022239235-A1
3          H02J1/10       True  False   []   US-2022239235-A1
4         H02J3/381       True  False   []   US-2022239235-A1
...             ...        ...    ...  ...                ...
92280  C02F2201/008      False  False   []     CN-113683210-A
92281     Y02W10/10      False  False   []     CN-113683210-A
92282      B64G4/00       True  False   []     CN-113682495-A
92283     B64G1/401       True  False   []     CN-113682495-A
92284    B64G1/1007       True   True   []     CN-113682495-A

[92285 rows x 5 columns]

In [28]:
# Inspect the datatype of each columns first element for cpc_data

for col in cpc_data.columns:
    print(
        col,
        ": ",
        type(cpc_data[col].iloc[0]),
        cpc_data[col].iloc[0],
    )

code :  <class 'str'> H02M7/5395
inventive :  <class 'numpy.bool_'> True
first :  <class 'numpy.bool_'> False
tree :  <class 'list'> []
publication_number :  <class 'str'> US-2022239235-A1


In [29]:
# Inspect if the tree columns list elements are empty

for i in range(len(cpc_data)):
    if len(cpc_data["tree"].iloc[i]) != 0:
        print(i)

Nothing is printed therefore we can safely drop the column tree from the cpc_data and then merge the cpc_data and cleantech_data on the "Publication_Number" column


In [30]:
# Print the shape of cleantech_patent_data and cpc_data
print(cpc_data.shape, cleantech_patent_data.shape)

print(
    "---------------------------------------------------------------------------",
)

# Drop the tree column

cpc_data.drop(["tree"], axis=1, inplace=True)

# Display the cleantech_patent_data dataframe
print(cpc_data.head())

# Merge the cpc_data and cleantech_patent_data dataframes

cleantech_patent_data = cleantech_patent_data.merge(cpc_data, on="publication_number")

print(
    "---------------------------------------------------------------------------",
)

# Display the cleantech_patent_data dataframe
print(cleantech_patent_data.head(2))

# Print the shale of all 3 dataframes to check if they are merged correctly
print(
    "---------------------------------------------------------------------------",
)
print(cpc_data.shape, cleantech_patent_data.shape)

# Print colnames of cleantech_patent_data
print(cleantech_patent_data.columns)

(92285, 5) (30000, 11)
---------------------------------------------------------------------------
         code  inventive  first publication_number
0  H02M7/5395       True  False   US-2022239235-A1
1    H02J3/32      False  False   US-2022239235-A1
2    H02M1/32       True  False   US-2022239235-A1
3    H02J1/10       True  False   US-2022239235-A1
4   H02J3/381       True  False   US-2022239235-A1
---------------------------------------------------------------------------
  publication_number application_number country_code  \
0   US-2022239235-A1  US-202217717397-A           US   
1   US-2022239235-A1  US-202217717397-A           US   

                                     title_localized  \
0  [{'text': 'Adaptable DC-AC Inverter Drive Syst...   
1  [{'text': 'Adaptable DC-AC Inverter Drive Syst...   

                                  abstract_localized publication_date  \
0  [{'text': 'Disclosed is an adaptable DC-AC inv...         20220728   
1  [{'text': 'Disclosed is an adapt

We see the successful merge from the results as the total columns are: 14


In [31]:
# Now remove the nested fields and the redundant columns

cleantech_patent_data.drop(
    ["title_localized", "abstract_localized", "inventor", "cpc"], axis=1, inplace=True
)

# Print cleantech_patent_data columns
print(cleantech_patent_data.columns)


# Display the cleantech_patent_data dataframe

cleantech_patent_data

Index(['publication_number', 'application_number', 'country_code',
       'publication_date', 'title', 'abstract', 'inventors', 'code',
       'inventive', 'first'],
      dtype='object')


publication_number application_number country_code publication_date  \
0        US-2022239235-A1  US-202217717397-A           US         20220728   
1        US-2022239235-A1  US-202217717397-A           US         20220728   
2        US-2022239235-A1  US-202217717397-A           US         20220728   
3        US-2022239235-A1  US-202217717397-A           US         20220728   
4        US-2022239235-A1  US-202217717397-A           US         20220728   
...                   ...                ...          ...              ...   
251034     CN-113682495-A  CN-202110997949-A           CN         20211123   
251035     CN-113682495-A  CN-202110997949-A           CN         20211123   
251036     CN-113682495-A  CN-202110997949-A           CN         20211123   
251037     CN-113682495-A  CN-202110997949-A           CN         20211123   
251038     CN-113682495-A  CN-202110997949-A           CN         20211123   

                                                    title  \
0       Adaptable DC-AC Inverter Drive System and Oper...   
1       Adaptable DC-AC Inverter Drive System and Oper...   
2       Adaptable DC-AC Inverter Drive System and Oper...   
3       Adaptable DC-AC Inverter Drive System and Oper...   
4       Adaptable DC-AC Inverter Drive System and Oper...   
...                                                   ...   
251034  Space film satellite integrating communication...   
251035  Space film satellite integrating communication...   
251036  Space film satellite integrating communication...   
251037  Space film satellite integrating communication...   
251038  Space film satellite integrating communication...   

                                                 abstract  \
0       Disclosed is an adaptable DC-AC inverter syste...   
1       Disclosed is an adaptable DC-AC inverter syste...   
2       Disclosed is an adaptable DC-AC inverter syste...   
3       Disclosed is an adaptable DC-AC inverter syste...   
4       Disclosed is an adaptable DC-AC inverter syste...   
...                                                   ...   
251034  The invention discloses a space film satellite...   
251035  The invention discloses a space film satellite...   
251036  The invention discloses a space film satellite...   
251037  The invention discloses a space film satellite...   
251038  The invention discloses a space film satellite...   

                                                inventors        code  \
0                                                      []  H02M7/5395   
1                                                      []    H02J3/32   
2                                                      []    H02M1/32   
3                                                      []    H02J1/10   
4                                                      []   H02J3/381   
...                                                   ...         ...   
251034  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...   B64G1/401   
251035  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...  B64G1/1007   
251036  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...    B64G4/00   
251037  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...   B64G1/401   
251038  [FENG XIANGCHAO, WANG PENGFEI, LI ZHEN, HE ZEP...  B64G1/1007   

        inventive  first  
0            True  False  
1           False  False  
2            True  False  
3            True  False  
4            True  False  
...           ...    ...  
251034       True  False  
251035       True   True  
251036       True  False  
251037       True  False  
251038       True   True  

[251039 rows x 10 columns]

Upon inspecting the clean_tech dataframe we see that the application number and publication number, rows 2 and 3 seems duplicate. We check for dupicates rows in the next step.


In [32]:
# Check for duplicates for each column as the there are nested data in some columns, which causes error when we use "cleantech_patent_data.duplicated().sum()"

for col in cleantech_patent_data.columns:
    print(col, ": ", cleantech_patent_data.duplicated(col).sum())

print(
    "---------------------------------------------------------------------------",
)

# Check for Uniques values in publication_number and application_number

print(
    "Unique values in publication_number: ",
    cleantech_patent_data["publication_number"].nunique(),
)
print(
    "Unique values in application_number: ",
    cleantech_patent_data["application_number"].nunique(),
)

publication_number :  242494
application_number :  242616
country_code :  251008
publication_date :  250876
title :  242757
abstract :  242607
inventors :  246031
code :  243942
inventive :  251037
first :  251037
---------------------------------------------------------------------------
Unique values in publication_number:  8545
Unique values in application_number:  8423


We expect publication number to be the unique attribute as it has more unique values, therefore we will drop duplicates rows from the dataframe using the "publication number" column.


In [33]:
# Drop the duplicates using publication_number

cleantech_patent_data.drop_duplicates(subset=["publication_number"], inplace=True)

# Reset the index to account for the dropped rows
cleantech_patent_data.reset_index(drop=True, inplace=True)

# Display the cleantech_patent_data dataframe
cleantech_patent_data

publication_number application_number country_code publication_date  \
0      US-2022239235-A1  US-202217717397-A           US         20220728   
1      US-2022239251-A1  US-202217580956-A           US         20220728   
2         EP-4033090-A1      EP-21152924-A           EP         20220727   
3        US-11396827-B2  US-202117606042-A           US         20220726   
4      US-2022228549-A1  US-202117153853-A           US         20220721   
...                 ...                ...          ...              ...   
8540     CN-214850044-U  CN-202120980740-U           CN         20211123   
8541     CN-214840532-U  CN-202121225269-U           CN         20211123   
8542     CN-214835218-U  CN-202121309411-U           CN         20211123   
8543     CN-113690940-A  CN-202111107806-A           CN         20211123   
8544     CN-113683210-A  CN-202110998828-A           CN         20211123   

                                                  title  \
0     Adaptable DC-AC Inverter Drive System and Oper...   
1     System for providing the energy from a single ...   
2         Verfahren zum steuern einer windenergieanlage   
3     Control method for optimizing solar-to-power e...   
4                Mutually supporting hydropower systems   
...                                                 ...   
8540                                     一种可调节太阳能板朝向的箱变   
8541  Garden waste anaerobic fermentation coupling m...   
8542                                     一种具有充电功能的无人机机库   
8543                                         一种风电供电控制方法   
8544                                一种基于无线控制的漂浮式太阳能曝气装置   

                                               abstract  \
0     Disclosed is an adaptable DC-AC inverter syste...   
1     In accordance with an example embodiment, a so...   
2     Verfahren zum Steuern einer Windenergieanlage ...   
3     A control method for optimizing a solar-to-pow...   
4     The mutually supporting hydropower systems inc...   
...                                                 ...   
8540  The utility model relates to a case becomes te...   
8541  The utility model discloses a gardens discarde...   
8542  The utility model relates to an unmanned air v...   
8543  The invention provides a wind power supply con...   
8544  本发明公开了一种基于无线控制的漂浮式太阳能曝气装置，包括漂浮板、太阳能转化机构以及输气泵，所...   

                                              inventors         code  \
0                                                    []   H02M7/5395   
1                                                    []    H02S40/38   
2     [Schaper, Ulf, von Aswege, Enno, Gerke Funcke,...   F03D7/0276   
3                                                    []    F24S50/00   
4                                     [CHEN, CHUN-HUEI]     F03B3/00   
...                                                 ...          ...   
8540                              [WANG HAI, WANG YANG]    Y02E10/50   
8541  [Gao Zhelu, Dai Jiangyue, LIU ZIYI, YAN CONGQI...    Y02E60/50   
8542  [ZHANG ANZHI, ZHANG WANYONG, HUANG HUIYONG, XU...    Y02E10/50   
8543         [ZHANG YUCHUAN, LONG HAIYANG, TAN CUNZHEN]    Y02E10/76   
8544                                        [An Zhixia]  C02F2209/22   

      inventive  first  
0          True  False  
1          True  False  
2          True   True  
3          True  False  
4          True  False  
...         ...    ...  
8540      False  False  
8541      False  False  
8542      False  False  
8543      False  False  
8544      False  False  

[8545 rows x 10 columns]

After dropping the duplicates, Now we have 8545 rows left, out of 30000 rows originally.


In [34]:
# Final Inspect the datatype of each columns of cleantech_patent_data's first element

for col in cleantech_patent_data.columns:
    print(
        col,
        ": ",
        type(cleantech_patent_data[col].iloc[0]),
        cleantech_patent_data[col].iloc[0],
    )

publication_number :  <class 'str'> US-2022239235-A1
application_number :  <class 'str'> US-202217717397-A
country_code :  <class 'str'> US
publication_date :  <class 'str'> 20220728
title :  <class 'str'> Adaptable DC-AC Inverter Drive System and Operation
abstract :  <class 'str'> Disclosed is an adaptable DC-AC inverter system and its operation. The system includes multiple DC input sources as input to provide a stable operation under various conditions. DC input sources may be added to the system or removed from the system without impacting the functionality of the system. The disclosed system is suited for solar energy harvesting in grid-connected or off-grid modes of operation.
inventors :  <class 'list'> []
code :  <class 'str'> H02M7/5395
inventive :  <class 'numpy.bool_'> True
first :  <class 'numpy.bool_'> False


We notice that the date format is string. We will need to convert this into a datetime object. We will use the "publication_date" column to do this.


In [35]:
# Convert publication_date from string to date

cleantech_patent_data["publication_date"] = pd.to_datetime(
    cleantech_patent_data["publication_date"], format="%Y%m%d"
)


# Print head of cleantech_patent_data['publication_date']

cleantech_patent_data["publication_date"].head()

0   2022-07-28
1   2022-07-28
2   2022-07-27
3   2022-07-26
4   2022-07-21
Name: publication_date, dtype: datetime64[ns]

### Transition Summary: Enhancing NLP Framework to Handle Multilingual Data

**Objective:**  
Due to the limitations of NLTK in handling our diverse multilingual data, which caused problems such as missing values and inefficient data processing, we decided to update our NLP tools to better support multiple languages.

**Challenges Identified with NLTK:**  
- Inadequate support for non-English languages, leading to data loss and incomplete preprocessing in texts such as Korean and Russian.
- Poor performance and scalability when managing large and complex datasets, affecting the overall efficiency of data processing.

**Proposed Solution:**  
To address these deficiencies, we adopted spaCy for its advanced linguistic features in English, Chinese, and Russian, and KoNLPy for specialized Korean text analysis.

**Implementation Strategy:**  
1. **Tool Installation:** Installed spaCy along with necessary language models and configured KoNLPy, ensuring compatibility with existing Java settings.
2. **Function Adaptation:** Updated the `preprocess_text` function to intelligently switch between spaCy and KoNLPy based on the detected language of the text.
3. **System Configuration:** Correctly set up `JAVA_HOME` and installed a 64-bit JVM to resolve any JVM compatibility issues.
4. **Verification and Rollout:** Conducted extensive tests to confirm performance gains and integrated the new setup into the existing NLP pipeline smoothly.

**Benefits Achieved:**  
- Significantly enhanced processing accuracy and efficiency for multilingual text.
- Improved scalability and performance, facilitating more robust data handling capabilities.

**Conclusion:**  
This strategic update not only resolved the issues with NLTK but also enhanced our NLP system's ability to handle multilingual datasets effectively, ensuring more reliable and accurate analysis moving forward.


In [36]:
# pip install spacy
# python -m spacy download en_core_web_sm
# python -m spacy download zh_core_web_sm
# python -m spacy download ru_core_news_sm
# pip install konlpy
# Java Virtual Machine (JVM) needed by konlpy

In [37]:
import spacy
from konlpy.tag import Okt  # Korean text processing


# Load spaCy models for English, Chinese, and Russian
nlp_en = spacy.load("en_core_web_sm")
nlp_zh = spacy.load("zh_core_web_sm")
nlp_ru = spacy.load("ru_core_news_sm")

# Initialize the Okt tokenizer for Korean
okt = Okt()


def preprocess_text(text, lang_code):
    # Select the appropriate NLP tool based on language
    if lang_code == "en":
        nlp = nlp_en
    elif lang_code == "zh":
        nlp = nlp_zh
    elif lang_code == "ru":
        nlp = nlp_ru
    elif lang_code == "ko":
        tokens = okt.morphs(text)
        return " ".join(tokens)
    else:
        return text  # If language not supported, return text as is

    # Process text using spaCy
    doc = nlp(text)
    tokens = [
        token.lemma_
        for token in doc
        if not token.is_stop and not token.is_punct and token.is_alpha
    ]
    return " ".join(tokens)

In [38]:
# Example usage within the DataFrame
cleantech_patent_data["title_preprocessed"] = cleantech_patent_data["title"].apply(
    lambda x: preprocess_text(
        x, "en"
    )  # Assuming 'en' or dynamically set this per row if you know the language
)
cleantech_patent_data["abstract_preprocessed"] = cleantech_patent_data[
    "abstract"
].apply(
    lambda x: preprocess_text(x, "en")  # Assuming 'en' or adjust as necessary
)

## 2. Text Preprocessing


# Preprocess 'title' and 'content' columns in the media dataset

cleantech_patent_data["title_preprocessed"] = cleantech_patent_data["title"].apply(
preprocess_text
)
cleantech_patent_data["abstract_preprocessed"] = cleantech_patent_data[
"abstract"
].apply(preprocess_text)


In [39]:
# Displaying a sample of the preprocessed data
print("Media Data Preprocessed Sample:")
cleantech_patent_data[["title_preprocessed", "abstract_preprocessed"]].head(10)

Media Data Preprocessed Sample:


title_preprocessed  \
0    adaptable DC AC Inverter Drive System Operation   
1  system provide energy single contiguous solar ...   
2      Verfahren zum steuern einer windenergieanlage   
3  control method optimize solar power efficiency...   
4                 mutually support hydropower system   
5                     system Method drive away goose   
6                                         clad sheet   
7               harden solar energy collector system   
8  system method hydro base electric power genera...   
9  system method remove dust solar panel surface ...   

                               abstract_preprocessed  
0  Disclosed adaptable DC AC inverter system oper...  
1  accordance example embodiment solar energy sys...  
2  Verfahren zum Steuern einer Windenergieanlage ...  
3  control method optimize solar power efficiency...  
4  mutually support hydropower system include hyd...  
5  system method drive geese away area employ pre...  
6  clad sheet roof wall clad sheet include mount ...  
7  hardened solar thermal energy collector STEC s...  
8  hydrodynamic power generation assembly method ...  
9  present system method waterless contactless sy...

In [40]:
# Check the data integrity of the preprocessed data

cleantech_patent_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8545 entries, 0 to 8544
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   publication_number     8545 non-null   object        
 1   application_number     8545 non-null   object        
 2   country_code           8545 non-null   object        
 3   publication_date       8545 non-null   datetime64[ns]
 4   title                  8545 non-null   object        
 5   abstract               8545 non-null   object        
 6   inventors              8545 non-null   object        
 7   code                   8545 non-null   object        
 8   inventive              8545 non-null   bool          
 9   first                  8545 non-null   bool          
 10  title_preprocessed     8545 non-null   object        
 11  abstract_preprocessed  8545 non-null   object        
dtypes: bool(2), datetime64[ns](1), object(9)
memory usage: 684.4+ 

## **Saving the Preprocessed Data**

After verifying that the data is clean and correctly formatted, it is crucial to save the processed data to a file for future use or further analysis.


In [41]:
# Save the DataFrame to CSV
cleantech_patent_data.to_csv("patent_preprocessed.csv", index=False, encoding="utf-8")